### PR TITLE
[Oneway][Part 4] Services can have multiple outbounds

### PR DESCRIPTION
**Pull request by AlexAPB**
Services are now mapped to multiple outbound types (one for each RPC type). This means that the Dispatcher now holds a map of service name to each service's outbounds. Consequently, channels hold multiple outbounds for services.

This modifies transport.Outbounds to 
````
// Outbounds encapsulates outbound types for a service
type Outbounds struct {
	Unary  UnaryOutbound
	Oneway OnewayOutbound
}
````

This creates the need for `yarpc.Outbounds` to inject into the dispatcher:
````
// Outbounds encapsulates a service and its outbounds
type Outbounds map[string]transport.Outbounds
````

### CODE REVIEW

File: dispatcher.go [hunk 1]
```diff
@@ -57,20 +57,22 @@ type Dispatcher interface {
...
 	Filter      transport.Filter
 	Interceptor transport.Interceptor
 
-	// TODO FallbackHandler for catch-all endpoints
-
 	Tracer opentracing.Tracer
 }
 
+// Outbounds encapsulates a service and its outbounds
+type Outbounds map[string]transport.Outbounds
```
Comment (abhinav) [inline dispatcher.go]: Yes! No more `transport.Outbounds` for users.

Comment (abhinav) [inline dispatcher.go]: For a separate diff: While we're at it, we may want to make `type Inbounds
[]transport.Inbound` in the `yarpc` package too. 

Created https://github.com/yarpc/yarpc-go/issues/414 to address this later.

File: dispatcher.go [hunk 2]
```diff
@@ -81,26 +83,57 @@ func NewDispatcher(cfg Config) Dispatcher {
...
 type dispatcher struct {
 	transport.Registry
 
-	Name        string
-	Outbounds   transport.Outbounds
+	Name string
+
+	inbounds  []transport.Inbound
+	outbounds Outbounds
+
```
Comment (abhinav) [inline dispatcher.go]: Drop the filter field if it's not being used anymore

File: dispatcher.go [hunk 3]
```diff
@@ -213,3 +245,26 @@ func (d dispatcher) Stop() error {
 
 	return nil
 }
+
+// getUniqueOutbounds helps ensure we only start/stop the same outbound once
+func (d dispatcher) getUniqueOutbounds() []transport.Outbound {
+	var unique []transport.Outbound
```
Comment (abhinav) [inline dispatcher.go]: This needs verifying but you can *probably* use a `map[transport.Outbound]` to dedupe.

File: dispatcher.go [hunk 4]
```diff
@@ -213,3 +245,26 @@ func (d dispatcher) Stop() error {
...
+		if outs.Unary == nil && outs.Oneway == nil {
+			continue
+		}
+
+		if outs.Unary == nil {
+			unique = append(unique, outs.Oneway)
+		} else if outs.Oneway == nil {
+			unique = append(unique, outs.Unary)
+		} else if outs.Unary.(transport.Outbound) == outs.Oneway.(transport.Outbound) {
+			unique = append(unique, outs.Unary)
```
Comment (abhinav) [inline dispatcher.go]: For this many if-else statements, you can use switch.

```
switch {
case outs.Unary == nil && outs.Oneway == nil:
  continue
case out.Oneway == nil:
    // ..
case out.Unary == nil:
    // ..
default:
  unique = append(unique, outs.Unary, outs.Oneway)
}
```

File: dispatcher_test.go [hunk 5]
```diff
@@ -35,6 +35,7 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"github.com/uber/tchannel-go"
+	"go.uber.org/yarpc"
```
Comment (abhinav) [inline dispatcher_test.go]: import group ordering

File: transport/channel.go [hunk 6]
```diff
@@ -44,20 +44,27 @@ type Channel interface {
 	// MAY be called multiple times for a request. The returned outbound MUST
 	// have already been started.
 	GetUnaryOutbound() UnaryOutbound
+	GetOnewayOutbound() OnewayOutbound
```
Comment (abhinav) [inline transport/channel.go]: Mention the expected behavior if this channel doesn't provide a oneway/unary outbound.

File: transport/outbound.go [hunk 7]
```diff
@@ -51,5 +51,22 @@ type UnaryOutbound interface {
...
 
-// Outbounds is a map of service name to Outbound for that service.
-type Outbounds map[string]UnaryOutbound
+// OnewayOutbound is a transport that knows how to send oneway requests for
+// procedure calls.
+type OnewayOutbound interface {
+	Outbound
+
+	// CallOneway sends the given request through this transport and returns its
+	// response.
```
Comment (abhinav) [inline transport/outbound.go]: It returns an ack, not a response.

File: dispatcher.go (hunk 1 above)
Comment (AlexAPB) [inline dispatcher.go]: :)

File: dispatcher.go (hunk 3 above)
Comment (AlexAPB) [inline dispatcher.go]: yup you're right, will modify with test

File: dispatcher.go [hunk 8]
```diff
@@ -230,3 +261,30 @@ func (d dispatcher) Stop() error {
...
+		}
+	}
+
+	for _, outs := range d.outbounds {
+		addIfUnique(outs.Unary)
+		addIfUnique(outs.Oneway)
+	}
+
+	unique := make([]transport.Outbound, 0, len(seen))
+	for outbound := range seen {
```
Comment (abhinav) [inline dispatcher.go]: Instead of doing this in two stages, just append to `unique` inside
`addIfUnique`.

File: transport/channel.go [hunk 9]
```diff
@@ -39,25 +39,33 @@ type Channel interface {
...
 }
 
-// IdentityChannel constructs a simple Channel for the given caller-service pair
-// which always returns the given Outbound.
-func IdentityChannel(caller, service string, out UnaryOutbound) Channel {
-	return identityChannel{caller: caller, service: service, outbound: out}
+// MultiOutboundChannel constructs a Channel backed by multiple outobund types
+func MultiOutboundChannel(caller, service string, Outbounds Outbounds) Channel {
+	return multiOutboundChannel{caller: caller, service: service, Outbounds: Outbounds}
```
Comment (abhinav) [inline transport/channel.go]: Let's remove this from the public API by moving it to an internal package.

File: dispatcher.go [hunk 10]
```diff
@@ -224,3 +256,27 @@ func (d dispatcher) Stop() error {
 
 	return nil
 }
+
+// getUniqueOutbounds helps ensure we only start/stop the same outbound once
+func (d dispatcher) getUniqueOutbounds() []transport.Outbound {
+	seen := make(map[transport.Outbound]struct{})
+	unique := []transport.Outbound{}
```
Comment (abhinav) [inline dispatcher.go]: Minor: You know the approximate size for both things: `len(d.outbounds)*2`. Use that in the `make` call.

```
seen := make(..., len(d.outbounds)*2)
unique := make([]transport.Outbound, 0, len(d.outbounds)*2)
```

**Comment by AlexAPB** [comment]
abandoned, applied diff in rebase (to be pushed)

**Pull request closed without merging**